# Commandline Usage


```bash
airrship [-h] [-v] -o OUT_NAME [--outdir OUTDIR] [--datadir DATADIR] 
         [-n NUMBER_SEQS] [--het PROP PROP PROP] [--shm] [--shm_multiplier SHM_MULTIPLIER] 
         [--shm_flat] [--mut_rate MUT_RATE | --mut_num MUT_NUM] [--shm_random] 
         [--all_alleles] [--locus LOCUS FILE] 
         [--flat_vdj {gene,family,False}] [--no_trim] 
         [--no_trim_v3] [--no_trim_d3] [--no_trim_d5] [--no_trim_j5] 
         [--no_np][--no_np1] [--no_np2] [--non_productive]
         [--prop_non_productive PROP] [--seed SEED] 
```

## Parameters

| Option                           | Details                                                                                                                                                                                                                                                                                                                                                                                                                  |
| -------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| -o, --outname &lt;out_prefix&gt; | Name for repertoire files. Only required parameter.                                                                                                                                                                                                                                                                                                                                                                      |
| --outdir &lt;out_dir&gt;         | Output directory. The current working directory is used if not specified.                                                                                                                                                                                                                                                                                                                                                |
| --datadir &lt;data_dir&gt;       | Alternative input data directory. Defaults to airrship/data. <br/>  Data must be formatted as in the airrship data directory                                                                                                                                                                                                                                                                                             |
| -n, --number_seqs &lt;n_seqs&gt; | Number of sequences to simulate. Defaults to 1000.                                                                                                                                                                                                                                                                                                                                                                       |
| --het &lt; prop prop prop &gt;   | Proportion of genes to be heterozygous, specify as V D J. <br/> Values must be between 0 and 1. Not compatible with --all_alleles.  <br/>  Not all genes  have  more than one allele. The proportion achieved may <br/> therefore be lower than requested. <br/> Defaults to 0.74 0.16 0.67 (the maximum possible proportion of  <br/> heterozygous genes using the included IMGT alleles).                              |
| --shm                            | Hypermutate sequences according to experimental parameters. <br/>  Each base will be mutated according to its 5mer context <br/> and the mutation frequency for each sequence will match observed <br/> distributions. If not specified, sequences will not be mutated. Mutation <br/> rates can be controlled by replacing the mut_freq_per_seq_per_family.csv <br/> reference file or specifying a --shm_multiplier. |
| --shm_multiplier                 | Multiplication factor to use on per sequence mutation rate distribution. <br/>Defaults to 1, i.e. replicates frequencies from the <br/> mut_freq_per_seq_per_family.csv reference file.  |
| --shm_flat                       | Mutate each sequence to the same degree (i.e. return a flat per <br/> sequence  mutation distribution). Specify degree of mutation <br/> using --mut_num or --mut_rate. Will default to a mutation rate of 0.05.                                                                                                                                                                                                         |
| --shm_random                     | Do not mutate individual bases according to kmer context. <br/> Each base will have an equal chance of being mutated.                                                                                                                                                                                                                                                                                                    |
| --mut_rate &lt;mut_rate&gt;      | Mutation frequency for flat SHM only. Value between 0 and 0.6. <br/> Not compatible with --mut_num. Defaults to 0.05.                                                                                                                                                                                                                                                                                                    |
| --mut_num &lt;number_muts&gt;    | Number of mutations for flat SHM only. <br/> Not compatible with --mut_rate.                                                                                                                                                                                                                                                                                                                                             |
| --all_alleles                    | Use all available alleles from all available genes, <br/> i.e., do not generate a synthetic 'haplotype'. Not compatible with --het.                                                                                                                                                                                                                                                                                      |
| --locus &lt;locus_file&gt;       | Do not generate a new locus, instead specify path <br/>  to an existing csv file to use as locus for repertoire generation.                                                                                                                                                                                                                                                                                              |
| --vdj_flat {gene, family}        | Do not use experimental data to bias VDJ usage, instead <br/> use all genes or families evenly.                                                                                                                                                                                                                                                                                                                          |
| --no_trim                        | Don't trim any end of any VDJ genes during recombination.                                                                                                                                                                                                                                                                                                                                                                |
| --no_trim_v3                     | Don't trim 3' end of V genes during recombination.                                                                                                                                                                                                                                                                                                                                                                       |
| --no_trim_d5                     | Don't trim 5' end of D genes during recombination.                                                                                                                                                                                                                                                                                                                                                                       |
| --no_trim_d3                     | Don't trim 3' end of D genes during recombination.                                                                                                                                                                                                                                                                                                                                                                       |
| --no_trim_j5                     | Don't trim 5' end of J genes during recombination.                                                                                                                                                                                                                                                                                                                                                                       |
| --no_np                          | Don't insert nucleotides at either gene junction, <br/> i.e., do not create NP regions.                                                                                                                                                                                                                                                                                                                                  |
| --no_np1                         | Don't insert nucleotides at the VD junction, <br/> i.e., do not create NP1 regions.                                                                                                                                                                                                                                                                                                                                      |
| --no_np2                         | Don't insert nucleotides at the DJ junction, <br/> i.e., do not create NP2 regions.                                                                                                                                                                                                                                                                                                                                      |
|--non_productive                 |  Include non-productive sequences in the output.<br/>This includes sequences with out of frame V and J segments,<br/> stop codons and/or missing junction anchor residues <br/>(C-104 and W/F-118). The majority of sequences produced will be<br/> non-productive (~75% using defaults without SHM, ~85% with SHM).<br/>  Specify --prop_non_productive to control this proportion.                                                                                                                                                                                                                                            |
| --prop_non_productive                 |  Proportion of sequences to be non-productive. <br/> Value between 0 and 1. Use with --non_productive.                                                                                                                                                                                                                                             |
| --seed &lt;seed&gt;              | Set random seed.                                                                                                                                                                                                                                                                                                                                                                                                         |